src/scmrepo/git/backend/dulwich/credentials.py

```diff
@@ -86,8 +86,9 @@ def _prepare_command(self) -> Union[str, List[str]]:
             ).strip()
             if shutil.which(executable, path=git_exec_path):
                 executable = os.path.join(git_exec_path, executable)
-
-        return [executable, *argv[1:]]
+        args = [executable]
+        args.extend(map(os.path.expanduser, argv[1:]))
```

Comment:
I'm not sure this is safe for us to do. It seems like the credential helper (in this case `git-credential-store`) should be responsible for doing this path expansion and not us, since it should really only be applied to file paths. There could be alternative credential helpers which use `~` in an argument (as a raw character in a username or password for example) and doing this expansion would break that functionality

Reply:
We may need to double check what cgit actually does here

Reply:
Looks like what's happened to me (and what motivated this PR) was that I set a _custom_ file for the credentials.
(caused by reading the git credentials-store docs in haste whilst simultaneously getting flummoxed with trying to get the dvc + git + docker combination working, (but that's another story....)

Seems there's 4 different ways to use git's credentials.helper feature:

1.  Default `~/.git-credentials` used, no URL context.
```bash
git config credentials.helper store
```
scmrepo will find credentials just fine from the default  `~/.git-credentials`  that is not explicitly stored in a  `.gitconfig ` or `.git/config` file.

2.  Custom file, no URL context.
```bash
git config credntials.helper "store --file ${CUSTOM_CREDS_FILE}"
```

3. Default file, with URL context,
```bash
git config credentials.${URL}.username ${USERNAME}
git config credentials.helper store
```

4. Custom file, with URL context.
```
git config credentials.${URL}.username ${USERNAME}
git config credentials.helper "store --file ${CUSTOM_CREDS_FILE}
```

`git config` adds `${CUSTOM_CREDS_FILE}`  _verbatim_  to `~/.gitconfig`  (or `~/.git/config` if  `git config --local` was used).

I believe only in the specific cases where method 2, 4 or if the user writes their .gitconfig by hand _and_ a the path specified starts with `~` does #178 apply where  no credentials are found by the code here.(as the path isn't interpolated in the stack anywhere),

So, in summary what I'd done is more-than-likely a minority issue that few others will run into, and most users using a `store` credential-helper, will probably(?) just use the credentials store command without explicitly using `--file`.

For what it's worth, seems from the git source[code](https://github.com/git/git/blob/c4203212e360b25a1c69467b5a8437d45a373cac/builtin/credential-store.c#L174) that _custom_ paths are appended to a list, and not interpolated.

If this PR warrants further consideration, I agree that the parsing could be more specific, e.g: split `argv[:1]` into tokens and only attempt to invoke `expanduser` on `argv[:1][2]`  in that sequence iif bot `argv[1:][0] == "store"` and  `argv[1:][1] == "--file"` (happy to  look into changing this PR to do that if there's interest).
 
Apologies if this has wasted your bandwidth @pmrowla , at least anyone else making the same mistake and falling down the ensuing rabbit hole will find this and go 🤦‍♂️ 😎 .

Reply:
Thanks for doing the investigation!

Given that cgit doesn't do the path expansion either, they probably just expect it to be handled by the shell. Would you mind testing a version of this PR that just changes the `subprocess.run` call to use `shell=True` (instead of using `expanduser()` on all of the command args)

Reply:
I implemented this locally, passes all tests.

Reply:
> Attempting to circumvent it, for example by converting the cmd list to a string and using shell=True triggers a CVE warning from bandit.

In this case the bandit warning is expected, and we could suppress it. (Yes, running an arbitrary user-provided shell command is dangerous, but it's also what cgit does in this situation)

Reply:
Ok, I took another look at the git source, and I think the existing DVC behavior is correct. CGit subcommands are run with `execv` (and not passed through the shell), and for git commands that support path expansion, they are looking for specific args that start with `~` and manually calling their `interpolate_path` function to do the expansion.

In the `git-credential-store` case, they do not do the interpolation (as you already noted), and pass whatever you set in `--file` directly into `fopen()`, so in the #178 case, it will look for a file in your current working dir named `./~/.git-credentials` (so it will look for a directory named `~` in your cwd).

https://github.com/git/git/blob/c4203212e360b25a1c69467b5a8437d45a373cac/builtin/credential-store.c#L173-L180
https://github.com/git/git/blob/c4203212e360b25a1c69467b5a8437d45a373cac/builtin/credential-store.c#L10-L20